subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java

```diff
@@ -1076,8 +1082,21 @@ private void rethrowFailure(String type, Collection<Throwable> failures) {
         throw new DefaultLenientConfiguration.ArtifactResolveException(type, getIdentityPath().toString(), getDisplayName(), failures);
     }
 
-
     private void assertResolvingAllowed() {
+        assertIsResolvable();
+
+        if (domainObjectContext.getProjectPath() != null) {
+            Project project = projectFinder.findProject(domainObjectContext.getProjectPath().getPath());
+            if (project != null) {
+                ProjectState projectState = projectStateRegistry.stateFor(project);
+                if (!projectState.hasMutableState()) {
+                    DeprecationLogger.nagUserOfDeprecatedBehaviour("The configuration " + identityPath.toString() + " was resolved without accessing the project in a safe manner.  This can happen because the configuration was resolved from a thread not managed by Gradle or because the configuration resides in a project other than the one that contains this task.");
```

Comment:
maybe make this less task specific?  Are there some ways to get here without using a task?

Reply:
I suppose someone could try to resolve a configuration in another project during configuration time, but this seems less likely than the other cases.  I was trying to avoid it being so abstract and vague that it wouldn't be meaningful and actionable to the user.  Another option might be to list the likely incorrect ways - something like:

```
This can happen because the configuration was resolved in one of the following scenarios:
 - From a thread not managed by Gradle 
 - From a task in a different project than the configuration's project
 - While evaluating a project other than the configuration's project
```

WDYT?

Reply:
Done.